synapse/util/caches/descriptors.py

```diff
@@ -444,7 +444,9 @@ def arg_to_cache_key(arg: Hashable) -> Hashable:
                     deferred: "defer.Deferred[Any]" = defer.Deferred()
                     deferreds_map[arg] = deferred
                     key = arg_to_cache_key(arg)
-                    cache.set(key, deferred, callback=invalidate_callback)
```

Comment:
this was the source of the problem.

`cache.set` creates a new observer, which we were then throwing away. So when the error happened, the failure would be propagated into the observer, but there was nothing to handle it.

One option was to change `cache.set` to avoid creating an observed, but that gets fiddly because it makes the other call to `cache.set` inconvenient/less efficient.

So instead we change things round and add the resulting observer to the list of deferreds we wait for down at line 479.